src/NuGet.Clients/NuGet.SolutionRestoreManager/ErrorListTableDataSource.cs

```diff
@@ -0,0 +1,183 @@
+                return _tableSubscription;
+            }
+        }
+
+        /// <summary>
+        /// Clear only nuget entries.
+        /// </summary>
+        public void ClearNuGetEntries()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            lock (_lockObj)
+            {
+                var sink = _tableSubscription?.TableDataSink;
+
+                if (sink != null)
+                {
+                    var entries = _errorList.TableControl?.Entries?.Where(IsNuGetEntry).ToArray()
+                        ?? new ITableEntryHandle[0];
+
+                    sink.RemoveEntries(entries);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Add error list entries
+        /// </summary>
+        public void AddEntries(params ErrorListTableEntry[] entries)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            lock (_lockObj)
+            {
+                var sink = _tableSubscription?.TableDataSink;
+
+                if (sink != null)
+                {
+                    sink.AddEntries(entries.ToList(), removeAllEntries: false);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Show error window.
+        /// </summary>
+        public void BringToFront()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            // Give the error list focus.
+            var vsErrorList = _errorList as IVsErrorList;
+            vsErrorList?.BringToFront();
+        }
+
+        private void EnsureInitialized()
+        {
+            if (!_initialized)
+            {
+                ThreadHelper.ThrowIfNotOnUIThread();
+
+                _errorList = _serviceProvider.GetService(typeof(SVsErrorList)) as IErrorList;
+                _tableManager = _errorList?.TableControl?.Manager;
+
+                if (_tableManager != null)
+                {
+                    _initialized = true;
+                    _tableManager.AddSource(this);
+                }
+            }
+        }
+
+        private static bool IsNuGetEntry(ITableEntryHandle entry)
+        {
+            object sourceObj;
+            return (entry != null
+                && entry.TryGetValue(StandardTableColumnDefinitions.ErrorSource, out sourceObj)
+                && StringComparer.Ordinal.Equals(ErrorListTableEntry.ErrorSouce, (sourceObj as string)));
+        }
+
+        public void Dispose()
+        {
+            if (_tableManager != null && _initialized)
+            {
+                try
+                {
+                    NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+                    {
+                        await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                        _tableManager.RemoveSource(this);
```

Comment:
I believe TableManager is free thread API. it doesn't need to be called from UI thread.

http://index/?leftProject=Microsoft.VisualStudio.Editor.Implementation&leftSymbol=l7bfsg1kywnl&file=TableManager%5CTableManager.cs